components/scream/src/physics/cld_fraction/cldfraction_main_impl.hpp

```diff
@@ -0,0 +1,88 @@
+#ifndef CLDFRACTION_MAIN_IMPL_HPP
+#define CLDFRACTION_MAIN_IMPL_HPP
+
+#include "physics/cld_fraction/cldfraction_functions.hpp"
+#include "ekat/kokkos/ekat_subview_utils.hpp"
+
+namespace scream {
+namespace cldfrac {
+
+/*-----------------------------------------------------------------*/
+template <typename S, typename D>
+void Functions<S,D>
+::cldfraction_main(
+  const Int nj,
+  const Int nk,
+  const view_2d<const Spack>& qi,
+  const view_2d<const Spack>& alst,
+  const view_2d<Spack>& aist,
+  const view_2d<Spack>& ast)
+{
+  using ExeSpace = typename KT::ExeSpace;
+  const Int nk_pack = ekat::npack<Spack>(nk);
+  const auto policy = ekat::ExeSpaceUtils<ExeSpace>::get_default_team_policy(nj, nk_pack);
+  Kokkos::parallel_for(
+    "cld fraction main loop",
+    policy,
+    KOKKOS_LAMBDA(const MemberType& team) {
+
+    const Int i = team.league_rank();
+
+    const auto oqi   = ekat::subview(qi,   i);
+    const auto oalst = ekat::subview(alst, i);
+    const auto oaist = ekat::subview(aist, i);
+    const auto oast  = ekat::subview(ast,  i);
+
+    cldfraction_calc_icefrac(team,nk,oqi,oaist);
+
+    cldfraction_calc_totalfrac(team,nk,oalst,oaist,oast);
+  });
+  Kokkos::fence();
+} // cldfraction_main
+/*-----------------------------------------------------------------*/
+template <typename S, typename D>
+KOKKOS_FUNCTION
+void Functions<S,D>
+::cldfraction_calc_icefrac(
+  const MemberType& team,
+  const Int& nk,
+  const uview_1d<const Spack>& qi,
+  const uview_1d<Spack>&       aist)
+{
+  team.team_barrier();
+  const Int nk_pack = ekat::npack<Spack>(nk);
+  Kokkos::parallel_for(
+    Kokkos::TeamThreadRange(team, nk_pack), [&] (Int k) {
+      auto icecld = qi(k) > 1e-5;
+      aist(k) = 0.0;
+      aist(k).set(icecld, 1.0);
+  }); // Kokkos_parallel_for nk_pack
+  team.team_barrier();
+} // cldfraction_calc_icefrac
+/*-----------------------------------------------------------------*/
+template <typename S, typename D>
+KOKKOS_FUNCTION
+void Functions<S,D>
+::cldfraction_calc_totalfrac(
+  const MemberType& team,
+  const Int& nk,
+  const uview_1d<const Spack>& alst,
+  const uview_1d<const Spack>& aist,
+  const uview_1d<Spack>&       ast)
+{
+  team.team_barrier();
+  const Int nk_pack = ekat::npack<Spack>(nk);
+  Kokkos::parallel_for(
+    Kokkos::TeamThreadRange(team, nk_pack), [&] (Int k) {
+      auto icecld = aist(k) > alst(k);
+      ast(k).set(Smask(true), alst(k));
```

Comment:
Smask should be true for cells we're actually doing calculations in and false for cells included only as padding, right? Or is it false where we do calculations? Smask is a confusing name. In any case, where is Smask calculated before it is used here?

Reply:
Smask is being constructed on demand here.  So `Smask(true)` passes a mask array of the appropriate size for the packed array `alst(k)`.  To answer your first question, this was my way to get `ast = max(alst,aist)`.  If there is a better way I'm happy to change it.  TBH, looking at this again, I think it would be just easier to set `ast(k)=alst(k)`, that should effectively copy the values of `alst` to `ast` and the mask properties, right?  << @bartgol

Reply:
Ok, your explanation make sense to me. And yeah, might as well ast(k)=alst(k) rather than using the mask.

Reply:
Two things:
- calling `a.set(Mask(true),b)` is the same as doing `a = b` (and might actually be better to avoid the mask, not sure);
- you can simply do `ast(k) = max(alst(k), aist(k)`, which should give you the same result as the current impl.

Reply:
This makes me wonder though: is `ast=max(aist,alst)`? I thought it was `ast=aist+alst`...

Reply:
Side note: the terms `Smask` and `Spack` always confused me. It's my Italian brain, probably: S-xyz in italian often means the opposite of xyz, so Smask sounds like 'unmask' to me... One day I'll propose for a different typedef. I haven't come up with one yet...

Reply:
I've made the change, `ast=max(aist,alst)` is much simpler.  For some reason I thought this would fail on a packed array.

Reply:
About my thumbs down emoji - I mean that it is max(alst,aist) rather than alst+aist... I realized after "reacting" that thumbs down might be construed as saying I thought it was a dumb question rather than reacting "NOOOO!!!!" like I meant to. :-)

Reply:
If that's the case, when we rename variables, I vote for `max_cld_frac` for the "total" one. Total always makes me thing of a sum.

Reply:
I'm not sure the total cloud fraction always needs to be the max of ice or liquid cloud fractions, so I don't like max_cld_frac. It can never be the sum of cld_frac_liq and cld_frac_ice, though, because if both liquid and ice cloud fractions were 100%, our total cloud fraction would be 200%! I guess it could be something like cld_frac_tot = min( cld_frac_liq+cld_frac_ice, 1) though. That would be "minimally overlapped" total cloud fraction...